### ISSUE TITLE
Don't crash when an unsupported method is requested

### ISSUE THREAD

**Issue by Seelengrab** [julia-vscode/LanguageServer.jl#1155]
It would be nice if LanguageServer.jl wouldn't crash if an unsupported method is requested by the editor. Got this while trying out `helix` and its LSP support:

```
2022-09-09T10:11:27.087 helix_lsp::transport [ERROR] err <- "ERROR: Unknown method textDocument/typeDefinition.\n"
2022-09-09T10:11:27.088 helix_lsp::transport [ERROR] err <- "Stacktrace:\n"
2022-09-09T10:11:27.108 helix_lsp::transport [ERROR] err <- " [1] error(s::String)\n"
2022-09-09T10:11:27.108 helix_lsp::transport [ERROR] err <- "   @ Base ./error.jl:35\n"
2022-09-09T10:11:27.109 helix_lsp::transport [ERROR] err <- " [2] dispatch_msg(x::JSONRPC.JSONRPCEndpoint{Base.PipeEndpoint, Base.PipeEndpoint}, dispatcher::JSONRPC.MsgDispatcher, msg::Dict{String, Any})\n"
2022-09-09T10:11:27.110 helix_lsp::transport [ERROR] err <- "   @ JSONRPC ~/.julia/packages/JSONRPC/P0G1p/src/typed.jl:81\n"
2022-09-09T10:11:27.110 helix_lsp::transport [ERROR] err <- " [3] run(server::LanguageServerInstance)\n"
2022-09-09T10:11:27.110 helix_lsp::transport [ERROR] err <- "   @ LanguageServer ~/.julia/packages/LanguageServer/0vsx2/src/languageserverinstance.jl:382\n"
2022-09-09T10:11:27.110 helix_lsp::transport [ERROR] err <- " [4] runserver(pipe_in::Base.PipeEndpoint, pipe_out::Base.PipeEndpoint, env_path::String, depot_path::String, err_handler::Nothing, symserver_store_path::Nothing)\n"
2022-09-09T10:11:27.110 helix_lsp::transport [ERROR] err <- "   @ LanguageServer ~/.julia/packages/LanguageServer/0vsx2/src/runserver.jl:41\n"
2022-09-09T10:11:27.111 helix_lsp::transport [ERROR] err <- " [5] runserver()\n"
2022-09-09T10:11:27.111 helix_lsp::transport [ERROR] err <- "   @ LanguageServer ~/.julia/packages/LanguageServer/0vsx2/src/runserver.jl:39\n"
2022-09-09T10:11:27.111 helix_lsp::transport [ERROR] err <- " [6] top-level scope\n"
2022-09-09T10:11:27.111 helix_lsp::transport [ERROR] err <- "   @ none:1\n"
2022-09-09T10:11:27.133 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-09-09T10:11:27.133 helix_lsp::transport [ERROR] err: <- StreamClosed
2022-09-09T10:11:45.200 helix_lsp::transport [ERROR] err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
2022-09-09T10:13:14.800 helix_lsp::transport [ERROR] err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
```

Having a log message is fine, having to restart the editor/language server is not :/

**Comment by davidanthoff** [comment]
That is a bug in helix' implementation: we report in our server capabilities that we do not support that method, so clients should not send that request.

In general, we _do_ want to crash for anything like that because we have crash reporting in the VS Code extension and that way the dev team learns about errors and can fix them.

**Comment by Seelengrab** [comment]
That makes sense, I did not know about this. That also makes the justification seem a bit off, since not everyone is going to run LS.jl under VSCode..

**Comment by pfitzseb** [comment]
> In general, we do want to crash for anything like that because we have crash reporting in the VS Code extension and that way the dev team learns about errors and can fix them.

We should presumably use the existing side-channel reporting that we already have. I'm personally in favor of moving us over to a never-crash implementation :)

**Comment by davidanthoff** [comment]
I have to admit I'm not a fan of making the LS more forgiving. I think in general for this kind of software, fail hard and fast if something is wrong is a much better design. And then fix any bugs that come up.

If we start to accept invalid requests, we'll have to come up with some line where we say "up to this point we accept invalid requests, but beyond that we think that things put us into an unstable state and there we stop". I really, really don't want to start wasting time on thinking those things through :)

> The client is free to send requests with any junk it wants to the server.

I kind of disagree with this. Just because the spec gives us the option to send an error message back, that does not mean that clients are free to send whatever request they want. In this case it seems very clear to me that `helix` should simply _not_ send this request because our server clearly indicates that we don't support it. Isn't that the whole point of the server and client capabilities design?

**Comment by Seelengrab** [comment]
I mean, if the end result is that I as a user have to potentially restart the LS over and over again due to an unwillingness to stay up, that's a really bad experience and would deter me from using it in the first place...

Theoretically speaking, `helix` or any other editor for that matter could also _send_ a correct message that got corrupted in transit somehow, would it be acceptable for LS to crash in that instance instead of even reporting the error and letting the user try again, potentially saving them having to restart the server & reindex their stuff?

**Comment by davidanthoff** [comment]
> I mean, if the end result is that I as a user have to potentially restart the LS over and over again due to an unwillingness to stay up, that's a really bad experience and would deter me from using it in the first place...

I think the right thing to do here is to lobby the `helix` devs to not send that message when a server indicates that it doesn't support it.

> Theoretically speaking, `helix` or any other editor for that matter could also send a correct message that got corrupted in transit somehow, would it be acceptable for LS to crash in that instance instead of even reporting the error and letting the user try again

I don't think we had any in-transit corruption ever that I'm aware of, based on the protocols we run over that just doesn't seem to be a thing. But, for arguments sake, let's say this is something that happens. Then, yes, there are definitely very many requests that would require us to shut down if they came in corrupted. Others maybe not, but figuring that out and making the LS robust with respect to that kind of error would add _a lot_ of work that I think we should spend on other things.

Here is an example: any of the notifications that change the state of the LS fall into the category that we need to shut down if we were to receive a corrupted message, no matter what. If the state of the documents in the client and the server gets out of sync (which we would have to assume to be the case if any of these messages get corrupted) nothing will work, and the LSP doesn't have any rules on how one could recover from such a situation. Shutdown really is the only sane option in that case.

There are other requests that would probably also fall into that category, i.e. pretty much anything that changes stuff. There might be other requests that are _not_ in that category where one could potentially ignore incorrect requests (generally things that purely request information), but thinking this kind of situation through and figuring out what exactly is safe and what isn't is a lot of work, hard to get right and in the end the better fix in a situation like this here is that the client gets fixed.

**Comment by Seelengrab** [comment]
> I think the right thing to do here is to lobby the helix devs to not send that message when a server indicates that it doesn't support it.

I have done exactly that already, as linked further up. Like I said though, this is irrespective of editor - presumably you'd run into the exact same issue should VSCode send a new LSP request to the server. I don't see how singling out `helix` here furthers the conversation, that just happened to be the editor I found this in.

> Then, yes, there are definitely very many requests that would require us to shut down if they came in corrupted. Others maybe not, but figuring that out and making the LS robust with respect to that kind of error would add a lot of work that I think we should spend on other things. [...]
> [...] but thinking this kind of situation through and figuring out what exactly is safe and what isn't is a lot of work, hard to get right and in the end the better fix in a situation like this here is that the client gets fixed.

Come on, that's kind of needless hyperbole... I'm not asking for arbitrary sanity and integrity checks on arbitrary messages, I'm literally just asking for "do we support this method? no? Let the client know so they can tell the user and continue business". The alternative is just that editors only report "The language server has crashed" when trying to use an unsupported method due to it crashing, which really just puts the _language server_ in a bad light. I don't think that's a particularly good look from a user perspective.

**Comment by jwortmann** [comment]
> I think in general for this kind of software, fail hard and fast if something is wrong is a much better design. And then fix any bugs that come up.

I kindly disagree with that. I think crashing as a matter of software design is only justifiable for a very particular type of software. Namely software which does some kind of calculations where the correctness of the results is critical for safety or if other important processes or decisions rely on the results. In that case, it is better to crash if there is any unexpected event, rather than to risk of having possibly incorrect results. But a language server is certainly not this kind of software. Language servers just exist to improve user experience in editors, and random crashing only makes this experience worse.

Especially relying on purposely crashing for your telemetry / error reporting is bad software design in my opinion. I'm not very familiar with the VSCode extension, but I would think that there must be another possible way to somehow intercept error messages from the server and communicate that to your issue reporting backend, if desired. That would at least be a better solution from the user perspective, and it would probably even reduce the risk of possibly losing unsaved work or something similar.

Besides that, every non-trivial software has bugs and in a constantly evolving system like the LSP, and where you have to communicate with plenty of other processes (clients) that are entirely out of your own control, you will probably never be able to take into account all possible contingencies of unexpected events.

> Just because the spec gives us the option to send an error message back, that does not mean that clients are free to send whatever request they want.

Clients probably *should* not do that, but apparently it does happen, and there is always the risk that another client will do something similar in the future, or interprets some part of the LSP specs (which are occasionally written in a imprecise way) a bit different than the server. But I think for this case it is quite clear that an error response would be most applicable according to the LSP specs. And if you start to question whether the specs should be followed or not - of course unless there is an obvious bug in the specs or something - then I'm pretty sure that cooperation with various clients will never become flawlessly :)

---

Don't get me wrong, I really appreciate all the work you're doing for the language server, the related packages & the Julia VSCode extension! And I'm aware that you're probably do it in your free time and don't get payed for it. But I have used several servers already and also work on the Sublime Text client regularly, and I think that there are some other servers which I have never seen crashing yet and which in my opinion provide a better user experience in regard to that matter.

**Comment by davidanthoff** [comment]
> I kindly disagree with that. I think crashing as a matter of software design is only justifiable for a very particular type of software. Namely software which does some kind of calculations where the correctness of the results is critical for safety or if other important processes or decisions rely on the results. In that case, it is better to crash if there is any unexpected event, rather than to risk of having possibly incorrect results. But a language server is certainly not this kind of software. Language servers just exist to improve user experience in editors, and random crashing only makes this experience worse.

The LS modifies the kind of code that you describe all the time. I can _very_ easily imagine a scenario where the LS has a corrupt state of the code internally for some reason, and then one of the features in the LS that modifies the code in the editor sends back a faulty (say old) version of the code to the editor (for example in a code formatting scenario). The user doesn't notice that the LS did not just format their code, but actually reverted it back to some outdated, previous version that contains a bug, runs their code, and now some numerical result was computed with faulty code because the LS was buggy.

Say the LSP adds a new notification in the future that is supposed to edit the LS state of the documents, _if_ a server opts into this via a server capability. Now imagine a client that doesn't correctly check server capabilities, and then sends this notification to a LS that actually doesn't support this particular notification. If we just ignored unknown messages, the LS would _not_ crash, but its internal state would be incorrect, and the scenario from the previous paragraph could easily occur.

**Comment by Seelengrab** [comment]
> The user doesn't notice that the LS did not just format their code, but actually reverted it back to some outdated, previous version that contains a bug, runs their code, and now some numerical result was computed with faulty code because the LS was buggy.
> [...]
> If we just ignored unknown messages, the LS would not crash, but its internal state would be incorrect, and the scenario from the previous paragraph could easily occur.

This only happens if the LS does _something_ on an unknown message. That's not what we're proposing.

> If we just ignored unknown messages, the LS would not crash, but its internal state would be incorrect, and the scenario from the previous paragraph could easily occur.

We're not talking about just ignoring the message though. It's perfectly fine to print a log message to let the user know and then not execute anything since the method is not supported after all (for example, I'd imagine this can be easily done using [window/showMessage](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showMessage)). However, outright crashing means there's no easy recourse for the user to continue working from a known good state (that being the one from before the client asked for a certain method), that may have taken quite a while to build up and only leads to frustration.

For julia in particular, crashing seems counterproductive - as I understand the architecture of LS.jl, restarting the LS means reindexing the symbol cache, which can be expensive. Letting the user know via `window/showMessage` just seems infinitely better in terms of UX, because it grants the user agency to do something else or just not use the unsupported feature.

**Comment by Seelengrab** [comment]
In regards to crashing, I also found this:

> if a client notices that a server exits unexpectedly, it should try to restart the server. However clients should be careful not to restart a crashing server endlessly. VS Code, for example, doesn’t restart a server which has crashed 5 times in the last 180 seconds.

So while I can see the appeal of crashing as a targettable metric, even for VS Code it seems suboptimal provided the language server crashes often & without user notification.. 

I'll check if there's an issue in the helix repo for tracking this too, to restart servers by themselves.

**Comment by davidanthoff** [comment]
> This only happens if the LS does something on an unknown message.

If the LS ignores a message that _should_ trigger an internal state change, then the LS has a corrupted internal state, and the situation is exactly identical. Just ignoring things does not mean one is on the safe side.

`window/showMessage` is guarded behind a client capability, so that is not always available. Also, in my mind sending a message a la "The LS might now be in a corrupt state, just so you know" is _really_ not helpful for the average user. What are they supposed to do at that point? If the LS crashes it is annoying, but at least at that point it is out of the way and the user doesn't have to worry about correctness issues.

> as I understand the architecture of LS.jl, restarting the LS means reindexing the symbol cache, which can be expensive.

No, that is not correct. Symbol caches are stored on disc, so nothing would have to be reindexed.

> even for VS Code it seems suboptimal provided the language server crashes often & without user notification..

VS Code _does_ notify users about this.

I think at this point it might be best if we just agree to disagree :) I won't add a functionality that blanket not-crashes on unknown messages.

Having said that, if someone wants to open a PR that for example adds an implementation for a specific request (say `textDocument/typeDefinition`) that sends back an error message instead of crashing, and that is for a known request that we can clearly say is not meant to modify the internal state of the LS, then I'd be happy to merge that if that is the only way to unblock say a scenario with helix. I still think it would be better to fix this in helix itself, but if that is not feasible for some reason, we could work around that on our side in the way just described.

**Comment by Seelengrab** [comment]
> If the LS ignores a message that should trigger an internal state change, then the LS has a corrupted internal state, and the situation is exactly identical. Just ignoring things does not mean one is on the safe side.

Again, that's not what I'm asking about. Please don't argue against a strawman. Noone is arguing for the LS not being allowed to try to recover.

> `window/showMessage` is guarded behind a client capability, so that is not always available. 

It being guarded behind a client capability is a bad argument against sending it when the client _may_ support that. Not wanting to do that just because the client may _not_ support it is not good behavior IMO. It's akin to not using LED lamps in traffic control because sometimes snow blocks them and LEDs don't put out enough heat to melt it.

> VS Code does notify users about this.

Not sure which part you're referring to - supporting `windows/showMessage` or notifying the user by itself that the LS crashed? 

> I won't add a functionality that blanket not-crashes on unknown messages.

Ok, then I think we're done here. I really hoped software quality & good UX was a higher priority for the sole language server for julia, but it seems that this is not the case.

This is a very fixable problem a real user & julia advocate encounters. Quite saddening to experience, I must say.

**Comment by davidanthoff** [comment]
> Again, that's not what I'm asking about. Please don't argue against a strawman. Noone is arguing for the LS not being allowed to try to recover.

Hm, maybe I am actually not understanding what you are suggesting... I thought you are suggesting that if the LS receives a message that it doesn't know about, it doesn't crash. How would the LS recover from that situation, if we don't even know what the semantics of the message is that we just didn't handle properly?

To make this concrete: let's say MS adds a new notification in some future iteration of the LSP `swapContentOfDoc`. The semantics of this notifications are that a LS should modify its internal representation of one of the files that is open in an editor. And let's assume that this feature is guarded behind a server capability. Now a (buggy) client comes along and starts to send this notification to servers that actually don't know about `swapContentOfDoc`. If the server now doesn't modify its internal state in response to a `swapContentOfDoc` notification, then the server has a corrupt state of the code files. If the user next sends a formatting request, the server will format the _old_, outdated version of the file content (which doesn't reflect the changes that the client sent via `swapContentOfDoc`) and send that back to the user. That seems like a _really_ bad situation to me.

What are the options here? One is your suggestion, to send a message to the user. What would that message say? "Hey, I received a message from the editor, don't know what it means, all bets are off, you better stop editing", or something like that? I genuinely can't think of a useful error message, other than saying "I give up and am out of the picture", which crashing is.

But if I misunderstood what you propose, please do explain it! I'm just not following right now.

> I really hoped software quality & good UX was a higher priority for the sole language server for julia, but it seems that this is not the case.

I would suggest that we abstain from accusations like this. I think I've explained quite clearly above that I do care about quality, but my priority is about correctness, and in my mind the suggestion on the table so far would compromise on that. If I misunderstood your proposal, I'm more than happy to listen to another round of explanations of what you are suggesting. 

I should also add that this really has _not_ been an issue for any even small number of users in many, many years. The vast majority of editors have never run into this kind of problem because they simply don't send messages that a server doesn't support based on the capabilities system. Now, I can understand that it can be frustrating if the editor of your choice has a problem, but I also made a very simple suggestion how you can fix this with a PR to the Julia LS that solves the crashes with `helix` while at the same time not compromising on the correctness concerns I have.

**Comment by Seelengrab** [comment]
> If the server now doesn't modify its internal state in response to a `swapContentOfDoc` notification, then the server has a corrupt state of the code files. If the user next sends a formatting request, the server will format the old, outdated version of the file content (which doesn't reflect the changes that the client sent via `swapContentOfDoc`) and send that back to the user. That seems like a really bad situation to me.
>
> What are the options here?

I'm not asking about notifications. I'm asking about message requests, which necessitate a response by the server, [as specified](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#requestMessage):

> Every processed request must send a response back to the sender of the request.

whose response should have `MethodNotFound` set [as its error](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#responseMessage):

> /**
>	 * The error object in case a request fails.
>	 */
>	error?: [ResponseError](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#responseError);


> export const MethodNotFound: [integer](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#integer) = -32601;

From this, I don't think it makes sense to crash when a request is received that the server does not support. 

For example, for ["Go To Type Definition"](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_definition) is a message request and as such the server must respond to it - even if only with an error message indicating `MethodNotFound`. The same is true for ["Go To Implementation"](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_implementation). 

I still don't follow how crashing instead of returning an error and carrying on can be a desirable behavior.

----

Of course, notifications are grouped into things the server [can ignore](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#dollarRequests) and those it can't - and for those it can't ignore its (as far as I can tell - I may have missed something in the docs) perfectly fine to crash if it doesn't support them, unless stated otherwise as is done for requests. I don't think its particularly useful to do so (after all, if the server knows it doesn't support a notification and thus _chooses_ to crash to force a restart, why can't it do a soft restart and start from its own entry point it would start from on a restart, without crashing?). But again, those are not what I'm asking about here.

**Comment by davidanthoff** [comment]
> as clients must be capable of handling a MethodNotFound if encountered.

I think at the end of the day this comes down to how much we want to rely on the client to do the right thing. Yes, I can imagine a scenario where a client doesn't check server capabilities, sends a request that the server doesn't support which is supposed to change the server state, then checks the return code of the request and realizes that the server didn't handle the request properly, and then it could maybe send some other request to try to trigger the same state change in the server or recover in some other clever way. But I can as easily imagine a client that ignores the return code and just continues as if nothing has gone wrong (after all, the client already ignored the server capabilities). On net, I just care more about correctness and making sure we never corrupt user code, and the easiest and safest way (in my view) is to shut down if we detect that a client is doing things we don't understand properly.

I just checked, there is actually a PR pending on helix to fix that on their side, so isn't this whole discussion going to be moot very soon in any case? As I wrote before, I'm not aware of any other editor/client that is running into this problem. Plus, if the helix fix is taking too long to be merged, do feel free to open a PR here that adds stub implementations for specific requests that send error messages back, that would be the simplest and quickest fix to get rid of the crashes you are seeing in any case.

> why can't it do a soft restart and start from its own entry point it would start from on a restart, without crashing?

There is no notion of a soft restart in the LSP, I don't see how that could be implemented.

**Comment by fredrikekre** [comment]
> There is no notion of a soft restart in the LSP, I don't see how that could be implemented.

In vim there is `:LspRestart` so if you are crashing the server you are back up again in 2 seconds. Probably something like that would be useful in helix too.

**Comment by davidanthoff** [comment]
@fredrikekre that terminates the LS process, and then starts the LS again, right?

I think @Seelengrab was thinking about something where the LS process would actually not terminate, but somehow reset its internal state and then start afresh, and I don't think that could be implemented. Completely agreed, an ability in the client to simply start/restart the LS is handy. I think VS Code actually doesn't have that... It just tries five times to restart the LS if it crashes and then gives up.

**Comment by fredrikekre** [comment]
Yes.

**Comment by Seelengrab** [comment]
> In vim there is :LspRestart so if you are crashing the server you are back up again in 2 seconds. Probably something like that would be useful in helix too.

Helix does have that on the master branch, but not in a release, and even VSCode does not have it, so why should other editors? Having such a command in the first place is a testament to not trusting language servers not to crash, it shouldn't be taken as a freebie to just crash because users can restart it anyway. 

> Plus, if the helix fix is taking too long to be merged, do feel free to open a PR here that adds stub implementations for specific requests that send error messages back, that would be the simplest and quickest fix to get rid of the crashes you are seeing in any case.

I have looked at the source code to try and do exactly that, but I honestly can't make sense of how the pieces fit together. I think you're talking about adding something to `messagedefs.jl` and also [here](https://github.com/julia-vscode/LanguageServer.jl/blob/9822394a1451e2ec726ad17d10e3fd8080d6ed58/src/languageserverinstance.jl#L330)? Such a stub seems to require structs and methods to call, at which point it seems I'd be closer to just implementing the features directly than having a "send error else do nothing" stub (I can't find it anymore, but there was some place I stumbled across in the code that explicitly had an `else` path that just used `error`, which I presume is where the error finally comes from - replacing this with `send_error` should be sufficient, I think.. not sure). Any pointers would be appreciated.

My position of "don't crash" is more or less founded on [this](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#lifeCycleMessages) part of the specification:

> The current protocol specification defines that the lifecycle of a server is managed by the client (e.g. a tool like VS Code or Emacs). It is up to the client to decide when to start (process-wise) and when to shutdown a server. 

I interpret that as "don't crash, the client will decide when you should stop". I acknowledge that you don't see it that way, but in terms of correctness/conforming to the spec, I don't think there's a (legal) way around this, irrespective of how lenient some clients are in respect to this lifecycle by being nice enough to restart the server for you :shrug:

**Comment by pfitzseb** [comment]
> Helix does have that on the master branch, but not in a release, and even VSCode does not have it, so why should other editors? 

VS Code definitely restarts the LS if it crashed (up to 5 times in as many minutes, I think).

Anyways, I'm sympathetic towards not crashing when idempotent requests are not handled/handle-able by the server -- it's impossible for those to corrupt internal state, so we might as well be nice about them and return an error code. "Just" needs someone to go through all requests and classify them :)

**Comment by Seelengrab** [comment]
> VS Code definitely restarts the LS if it crashed (up to 5 times in as many minutes, I think).

I was referring to the manual restart functionality that was mentioned above. The LSP docs mention that VSCode restarts a crashing LS up to 5 times in the last 180s - if it happens more often, it gives up.

> "Just" needs someone to go through all requests and classify them :)

I've started to look into this and I've decided that my first course of action should be documenting/cleaning up JSONRPC.jl - seems like there are [some LSP specific bits](https://github.com/julia-vscode/JSONRPC.jl/issues/68) in there that should really be handled here, instead of in a general server implementation. I've found that while trying to understand how this package works internally, so might as well do that while I'm at it.
